python/KMS_KEYS_TO_NOT_DELETE.py

```diff
@@ -0,0 +1,452 @@
+	Then: Return NON_COMPLIANT
+
+	Scenario 5:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The KMS Key is not one of the keys in the parameter
+	Then: Return NOT_APPLICABLE
+
+	Scenario 6:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The CMK is one of the keys in the parameter
+	And: The CMK is not scheduled for deletion
+	Then: Return COMPLIANT
+
+	Scenario 7:
+	Given: At least 1 CMK is present in the AWS Account
+	And: Rule parameter kmsKeyIds are configured and valid
+	And: The CMK is one of the keys in the parameter
+	And: The CMK is scheduled for deletion
+	Then: Return NON_COMPLIANT
+'''
+
+import json
+import sys
+import datetime
+import re
+import boto3
+import botocore
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::KMS::Key'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    evaluations = []
+    kms_client = get_client('kms', event)
+    account_keys = get_account_keys(kms_client)
+
+    if not account_keys:
+        return None
+
+    if 'kmsKeyIds' in valid_rule_parameters:
+        for key_id in valid_rule_parameters['kmsKeyIds']:
+            if key_id in account_keys:
+                kms_key_details = kms_client.describe_key(KeyId=key_id)
+                if kms_key_details['KeyMetadata']['Origin'] == 'AWS_KMS' and kms_key_details['KeyMetadata']['KeyManager'] == 'CUSTOMER' and kms_key_details['KeyMetadata']['KeyState'] == 'PendingDeletion':
+                    evaluations.append(build_evaluation(key_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='The KMS Key is scheduled for deletion.'))
+                    continue
+                evaluations.append(build_evaluation(key_id, 'COMPLIANT', event))
+        return evaluations
+
+    for key_id in account_keys:
+        kms_key_details = kms_client.describe_key(KeyId=key_id)
+        if kms_key_details['KeyMetadata']['Origin'] == 'AWS_KMS' and kms_key_details['KeyMetadata']['KeyManager'] == 'CUSTOMER' and kms_key_details['KeyMetadata']['KeyState'] == 'PendingDeletion':
+            evaluations.append(build_evaluation(key_id, 'NON_COMPLIANT', event, DEFAULT_RESOURCE_TYPE, annotation='The KMS Key is scheduled for deletion.'))
+            continue
+        evaluations.append(build_evaluation(key_id, 'COMPLIANT', event))
+    return evaluations
+
+def get_account_keys(kms_client):
+    account_key_library = []
+    response = kms_client.list_keys()
```

Comment:
Add limit to 1000: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/kms.html#KMS.Client.list_keys